R/predict-support.R

```diff
@@ -0,0 +1,188 @@
+##' Create parameters that include beta values correspond to different
+##' Rt scenarios in the future.
+##'
+##' This function is called `add_future_betas` because updates beta
+##' values, however it does this by applying Rt changes in the future
+##' via `future_Rt`.
+##'
+##' There are two basic ways that new beta values in the future might
+##' be added. First, as relative to some previous day in the past. For
+##' example to use a value of beta that corresponds to 1.5x the Rt
+##' recorded on the 28th of October, we might write
+##'
+##' ```
+##' sircovid::future_Rt(1.5, "2020-10-28")
+##' ```
+##'
+##' Second, as an absolute Rt value. For example to use a value of
+##' beta corresponding to an Rt of 0.9 we might write
+##'
+##' ```
+##' sircovid::future_Rt(0.9)
+##' ```
+##'
+##' These new levels need to be associated with dates at which they
+##' occur, and will result in a piece-wise constant pattern with 1
+##' day's transition between each level.
+##'
+##' @title Create future betas
+##'
+##' @param sample A `mcstate_pmcmc` object
+##'
+##' @param rt The results of [sircovid::carehomes_Rt_trajectories]
+##'   with matrix elements `Rt_general` and `date` which will be used to
+##'   calculate relative Rt values.
+##'
+##' @param future A named list of [sircovid::future_Rt] values. Each
+##'   element name corresponds to the date that the change takes
+##'   effect (in ISO-8601 "YYYY-MM-DD" format) and must be strictly
+##'   increasing with at least two days separating changes.
+##'
+##' @export
+##' @author Richard Fitzjohn
+add_future_betas <- function(sample, rt, future) {
+  assert_is(sample, "mcstate_pmcmc")
+  assert_is(rt, "Rt_trajectories")
+  sample <- drop_trajectory_predicted(sample)
+
+  ## We never want the future projections here, so exclude them
+  n_pars <- nrow(sample$pars)
+  n_time <- length(sample$trajectories$step)
+
+  ## Check that our trajectories are consistent with the sample:
+  i <- match(sample$trajectories$step, rt$step[, 1])
```

Comment:
This check is causing problems when I integrate it into the task, specifically we have
>   range(sample$trajectories$date)
[1]   1 389
>   range(sample_Rt$date)
[1]  17 389 ` 

because `sample_Rt` starts from the date of importation, whereas `sample` always starts from 2020-01-01, regardless of the date of importation (with corresponding NA entries in state). 
There are a few ways we could fix this:
1. change the Rt calc to match the output of sample (a bit of a pain as we would then need full pmcmc reruns to do anything downstream) 
2. remove the pre-importation dates (and states) from sample$trajectories, which is probably worth doing in another helper function a là `drop_trajectory_predicted()`
3. something I haven't thought of